File: tests/fixtures/project_ts/src/users/queries.ts
/**
 * User database queries.
 */

import { db } from "../db/client";
import { User, CreateUserInput, UpdateUserInput } from "./types";
import { NotFoundError } from "../utils/errors";
import { hashPassword } from "../utils/crypto";

export async function getUserById(userId: number): Promise<User> {
  const result = await db.query<User>("SELECT * FROM users WHERE id = $1", [userId]);
  if (result.rows.length === 0) {
    throw new NotFoundError("User", userId);
  }
  return result.rows[0];
}

export async function getUserByEmail(email: string): Promise<User | null> {
  const result = await db.query<User>("SELECT * FROM users WHERE email = $1", [email]);
  return result.rows.length > 0 ? result.rows[0] : null;
}

export async function createUser(input: CreateUserInput): Promise<User> {
  const passwordHash = await hashPassword(input.password);
  const result = await db.query<User>(
    "INSERT INTO users (email, name, password_hash, created_at, updated_at, is_active) VALUES ($1, $2, $3, NOW(), NOW(), true) RETURNING *",
    [input.email, input.name, passwordHash]
  );
  return result.rows[0];
}

export async function updateUser(userId: number, input: UpdateUserInput): Promise<User> {
  const existing = await getUserById(userId);

  const updates: string[] = [];
  const values: unknown[] = [];
  let paramIndex = 1;

  if (input.email !== undefined) {
    updates.push(`email = $${paramIndex++}`);
    values.push(input.email);
  }
  if (input.name !== undefined) {
    updates.push(`name = $${paramIndex++}`);
    values.push(input.name);
  }
  if (input.password !== undefined) {
    const newHash = await hashPassword(input.password);
    updates.push(`password_hash = $${paramIndex++}`);
    values.push(newHash);
  }
  if (input.isActive !== undefined) {
    updates.push(`is_active = $${paramIndex++}`);
    values.push(input.isActive);
  }

  if (updates.length === 0) {
    return existing;
  }

  updates.push(`updated_at = NOW()`);
  values.push(userId);

  const result = await db.query<User>(
    `UPDATE users SET ${updates.join(", ")} WHERE id = $${paramIndex} RETURNING *`,
    values
  );
  return result.rows[0];
}

export async function deleteUser(userId: number): Promise<void> {
  await getUserById(userId); // throws NotFoundError if not exists
  await db.query("DELETE FROM users WHERE id = $1", [userId]);
}

export async function listUsers(page: number = 1, limit: number = 20): Promise<User[]> {
  const offset = (page - 1) * limit;
  const result = await db.query<User>("SELECT * FROM users ORDER BY created_at DESC LIMIT $1 OFFSET $2", [
    limit,
    offset,
  ]);
  return result.rows;
}

File: tests/fixtures/project_py/src/users/queries.py
from __future__ import annotations

from src.db.client import db
from src.users.types import CreateUserInput, UpdateUserInput, User
from src.utils.crypto import hash_password
from src.utils.errors import NotFoundError


def get_user_by_id(user_id: int) -> User:
    """Fetch a user by their ID. Raises NotFoundError if not found."""
    result = db.query("SELECT * FROM users WHERE id = ?", (user_id,))
    if not result.rows:
        raise NotFoundError(resource="User", identifier=user_id)
    return User(**result.rows[0])


def create_user(data: CreateUserInput) -> User:
    """Create a new user from validated input."""
    hashed, salt = hash_password(data.password)
    db.execute(
        "INSERT INTO users (email, name, hashed_password, salt) VALUES (?, ?, ?, ?)",
        (data.email, data.name, hashed, salt.hex()),
    )
    result = db.query("SELECT * FROM users WHERE email = ?", (data.email,))
    return User(**result.rows[0])


def update_user(user_id: int, data: UpdateUserInput) -> User:
    """Update an existing user. Raises NotFoundError if the user does not exist."""
    get_user_by_id(user_id)  # raises NotFoundError if not found

    updates: dict[str, str | bool | None] = {}
    if data.email is not None:
        updates["email"] = data.email
    if data.name is not None:
        updates["name"] = data.name
    if data.password is not None:
        hashed, _salt = hash_password(data.password)
        updates["hashed_password"] = hashed
    if data.is_active is not None:
        updates["is_active"] = data.is_active

    if updates:
        set_clause = ", ".join(f"{k} = ?" for k in updates)
        db.execute(
            f"UPDATE users SET {set_clause} WHERE id = ?",
            (*updates.values(), user_id),
        )

    return get_user_by_id(user_id)


def delete_user(user_id: int) -> None:
    """Delete a user by ID. Raises NotFoundError if the user does not exist."""
    _ = get_user_by_id(user_id)  # ensure exists
    db.execute("DELETE FROM users WHERE id = ?", (user_id,))
